File: server/state.py
import json
import os

filename = 'state.json'

def joinObjects(obj1, obj2, deep):
    #TODO: add deep merge
    for key in obj2:
        obj1[key] = obj2[key]

def isObject(dict1, key):
    return type(dict1[key]) is dict #key in dict1 and 
    #TODO: lists should work as well
  
def isNone(obj, key):
    if type(obj) is dict:
        return not key in obj
    else:
        return key >= len(obj)
        
def follow(dest, path):
    for i in range(len(path) - 1):
        if isNone(dest, path[i]):
            dest[path[i]] = {} if type(path[i+1]) is str else []
        dest = dest[path[i]]
    return dest

class StateManager:
    def __init__(self):
        if os.path.exists(filename):
            with open(filename, 'r') as f:
                self.state = json.loads(f.read())
        else:
            self.state = {'curContext':[], 'mainContext':{'canvas': {'pos':[0,0], 'scale':1}}, 'routine':[], 'idCounter':1}
            with open(filename, 'w') as out:
                out.write(json.dumps(self.state))
        self.loadContext()

    def followPath(self, path):
        dest = self.state if path[0] == 'state' else self.context
        return follow(dest, path[1:])
        
    def commit(self, change):
        print(change)
        print('')
        dest = None
        last = change['path'][-1]
        value = change['value'] if 'value' in change else None
        
        if 'sourceMode' in change:
            if len(change['value']) > len(change['path']) and change['mode'] == 'insert':
                dest = self.followPath(change['path'])
                dest.insert(last, value)
                change['mode'] = 'replace'
            src = self.followPath(change['value'])
            srcKey = change['value'][-1]
            value = src[srcKey]
            if 'sourceMode' in change and change['sourceMode'] == 'move':
                del src[srcKey]
                
        dest = dest or self.followPath(change['path'])

        if change['mode'] == 'delete':
            del dest[last]
        elif change['mode'] == 'insert':
            dest.insert(last, value)
        elif change['mode'] == 'merge' and not isNone(dest, last) and isObject(dest, last):
            joinObjects(dest[last], value, False)
        else:
            dest[last] = value
         
        with open('state.json', 'w') as out:
            out.write(json.dumps(self.state))

    def loadContext(self):
        #print('Context loaded.')
        self.context = self.state['mainContext']
        for id in self.state['curContext']:
            self.context = self.context['containers'][id]['inner']

    def switchContext(self, newContext):
        self.state['curContext'] = newContext
        self.loadContext()
        with open('state.json', 'w') as out:
            out.write(json.dumps(self.state))

    def jumpIn(self, id):
        self.state['curContext'].append(id)
        self.context = self.context['containers'][id]['inner']

    def jumpOut(self):
        id = self.state['curContext'].pop()
        self.loadContext()
        return id

File: state.js
/* type, body

   type: execute, body:
        id: output element id

   type: contextSwitch, body:
        newContext: list of context path

   type: update, body:
        path: list of dict keys
        value: new value
        delete: boolean
        strict: boolean, replace old value, otherwise merge
        deep: boolean, merge deeply
*
*
*   changes: start path with context/state
        mode: 'replace', 'merge', 'deep merge', 'insert', 'delete'
*       sourceMode: 'copy', 'move', undefined
*
* */
/*
class Foo {
  constructor(msg) {

    if (Foo.instance) {
      return Foo.instance;
    }
    Foo.instance = this;
  }
}
    // Properties & Methods
}*/

import {CanvasManager} from "./canvas.js";
import {RoutineManager} from "./routines.js";
import {FakeServer, FakeHTTP} from "./fake_server.js";

const fakeServer = new FakeServer();

const canvasM = new CanvasManager();

const routineM = new RoutineManager();


const isObject = function(obj) {
    return typeof obj === 'object' && !!obj;
};

const joinObjects = function(obj1, obj2, deep){
    // TODO: add deep merge
    for (var attrname in obj2) { obj1[attrname] = obj2[attrname];}
};

export const followPath = function (dict, path) {
    for (var i = 0; i < path.length - 1; i++) {
        if (!isObject(dict[path[i]])){
            if(typeof path[i+1] === 'string'){
                dict[path[i]] = {};
            }else{
                dict[path[i]] = [];
            }
        }
        dict = dict[path[i]];
    }
    return dict;
};


export class StateManager{
    constructor(){
        if (StateManager.instance) {
            return StateManager.instance;
        }
        StateManager.instance = this;

        var jsonFile = new XMLHttpRequest();
        jsonFile.open("GET",document.URL+"state.json?"+ new Date().getTime(),true);


        this.xhttp = new FakeHTTP();

        jsonFile.onreadystatechange = function() {
            if (jsonFile.readyState== 4 && jsonFile.status == 200) {
                console.log('JSON loaded');
                this.state = JSON.parse(jsonFile.responseText);
                fakeServer.init(jsonFile.responseText);
                jsPlumb.ready(function () {
                    routineM.init();
                    canvasM.init();
                    this.loadContext();
                    routineM.loadRoutine();
                }.bind(this));

                // routineCallback();
            }
        }.bind(this);
        jsonFile.send();

    }

    loadContext(){
        this.context = this.state.mainContext;
        for (var i = 0; i < this.state.curContext.length; i++) {
            this.context = this.context.containers[this.state.curContext[i]].inner;
        }
        canvasM.loadContext();
    }

    switchDown(id){
        this.state.curContext.push(id);
        this.switchContext(this.state.curContext);
    }

    switchUp(){
        this.state.curContext.pop();
        this.switchContext(this.state.curContext);
    }

    switchContext(newContext){
        var exhttp = new FakeHTTP();
        exhttp.open("POST", document.URL, true);
        exhttp.send(JSON.stringify({type:"contextSwitch", body: newContext}));
        this.state.curContext = newContext;
        this.loadContext();
    }

    commitChange(change) {

        if(!change.mode){
            change.mode = 'merge';
        }

        console.log(change);

        //queue and send json, flush queue
        this.xhttp.open("POST", document.URL, true);
        this.xhttp.send(JSON.stringify({type:"update", body: change}));

        var dest = undefined;
        //console.log(change.path[change.path.length-1]);
        var last = change.path[change.path.length-1];
        
        var value = change.value;

        if(change.sourceMode){
            if(change.value.length > change.path.length && change.mode === 'insert'){
                dest = followPath(this, change.path);
                dest.splice(last, 0, undefined);
                change.mode = 'replace';
            }
            var src = followPath(this, change.value);
            var srcKey = change.value[change.value.length-1];
            value = src[srcKey];
            if(change.sourceMode === 'move'){
                if(Array.isArray(src)){
                    src.splice(srcKey, 1);
                }else{
                    delete src[srcKey];
                }
            }
        }

        if(!dest){
            dest = followPath(this, change.path);
        }

        switch (change.mode) {
            case 'delete':
                if(Array.isArray(dest)){
                    dest.splice(last, 1);
                }else{
                    delete dest[last];
                }
                break;
            case 'merge':
                if(isObject(dest[last])){ //TODO: handle list merge
                    joinObjects(dest[last], value, false);
                    break;
                }
            case 'replace':
                dest[last] = value;
                break;
            case 'insert':
                dest.splice(last, 0, value);
                break;
            default:
                console.error('Used unknown commit mode.');

        }




    }


}

new StateManager();
